### PR TITLE
[Merged by Bors] - feat: vertex replacement

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
`cliqueFree_of_replaceVertex_cliqueFree` is still quite long.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Basic.lean [hunk 1]
```diff
@@ -1216,6 +1216,47 @@ theorem DeleteFar.mono (h : G.DeleteFar p r₂) (hr : r₁ ≤ r₂) : G.DeleteF
 
 end DeleteFar
 
+/-! ## Vertex replacement -/
+
+
+/-- The graph formed by replacing `t` with a copy of `s`, by changing `t`'s neighbours to match.
+The `s-t` edge is removed if present. -/
+def replaceVertex (s t : V) : SimpleGraph V where
+  Adj := G.Adj \ Sym2.ToRel (G.incidenceSet t) ⊔ Sym2.ToRel {⟦(t, v)⟧ | v ∈ G.neighborSet s \ {t}}
```
Comment (eric-wieser) [inline Mathlib/Combinatorics/SimpleGraph/Basic.lean:1225]: You could do
```lean
Adj u v :=
  if u = t then
    If v = t then
        False
    else
        G.Adj s v
  else
    If v = t then
        G.Adj u s
    else
        G.Adj u v
```
If you're lucky the proofs follow by case-bashing, but I have no idea if this is bettee than your approach.

Comment (eric-wieser) [reply Mathlib/Combinatorics/SimpleGraph/Basic.lean:1225]: Ah, that's not quite right because I didn't handle the s-t edge

Comment (Parcly-Taxel) [reply Mathlib/Combinatorics/SimpleGraph/Basic.lean]: Actually your if-else tree works once `[DecidableEq V]` is added – see latest commit – and all proofs shorten with this change. But `cliqueFree_of_replaceVertex_cliqueFree` is still quite long.

File: Mathlib/Combinatorics/SimpleGraph/Basic.lean [hunk 2]
```diff
@@ -1216,6 +1216,28 @@ theorem DeleteFar.mono (h : G.DeleteFar p r₂) (hr : r₁ ≤ r₂) : G.DeleteF
...
+
+section ReplaceVertex
+
+variable [DecidableEq V] (s t : V)
+
+/-- The graph formed by replacing `t` with a copy of `s`, by changing `t`'s neighbours to match.
+The `s-t` edge is removed if present. -/
+abbrev replaceVertex : SimpleGraph V where
+  Adj v w := if v = t then if w = t then False else G.Adj s w
+                      else if w = t then G.Adj s v else G.Adj v w
```
Comment (ocfnash) [inline Mathlib/Combinatorics/SimpleGraph/Basic.lean:1230]: I think this is slightly clearer:
```suggestion
                      else if w = t then G.Adj v s else G.Adj v w
```

File: Mathlib/Combinatorics/SimpleGraph/Basic.lean [hunk 3]
```diff
@@ -1216,6 +1216,28 @@ theorem DeleteFar.mono (h : G.DeleteFar p r₂) (hr : r₁ ≤ r₂) : G.DeleteF
...
 
+/-! ## Vertex replacement -/
+
+
+section ReplaceVertex
+
+variable [DecidableEq V] (s t : V)
+
+/-- The graph formed by replacing `t` with a copy of `s`, by changing `t`'s neighbours to match.
+The `s-t` edge is removed if present. -/
```
Comment (ocfnash) [inline Mathlib/Combinatorics/SimpleGraph/Basic.lean:1227]:
```suggestion
/-- The graph formed by forgetting `t`'s neighbours and instead giving it those of `s`. The `s-t`
edge is removed if present. -/
```

File: Mathlib/Combinatorics/SimpleGraph/Basic.lean [hunk 4]
```diff
@@ -1216,6 +1216,28 @@ theorem DeleteFar.mono (h : G.DeleteFar p r₂) (hr : r₁ ≤ r₂) : G.DeleteF
...
+  Adj v w := if v = t then if w = t then False else G.Adj s w
+                      else if w = t then G.Adj s v else G.Adj v w
+  symm v w := by dsimp only; split_ifs <;> simp [adj_comm]
+
+theorem not_adj_of_replaceVertex : ¬(G.replaceVertex s t).Adj s t := by simp
+
+@[simp]
+theorem replaceVertex_self : G.replaceVertex s s = G := by
+  ext; dsimp only; split_ifs <;> simp_all [adj_comm]
+
```
Comment (ocfnash) [inline Mathlib/Combinatorics/SimpleGraph/Basic.lean:1238]: We should characterise `replaceVertex` with some API:
```suggestion

/-- There is never an `s-t` edge in `G.replaceVertex s t`. -/
lemma not_adj_replaceVertex_same :
    ¬ (G.replaceVertex s t).Adj s t := by
  simp

/-- Except possibly for `t`, the neighbours of `s` in `G.replaceVertex s t` are its neighbours in
`G`. -/
lemma adj_replaceVertex_iff_of_ne_left {w : V} (hw : w ≠ t) :
    (G.replaceVertex s t).Adj s w ↔ G.Adj s w := by
  simp [hw]

/-- Except possibly for itself, the neighbours of `t` in `G.replaceVertex s t` are the neighbours of
`s` in `G`. -/
lemma adj_replaceVertex_iff_of_ne_right {w : V} (hw : w ≠ t) :
    (G.replaceVertex s t).Adj t w ↔ G.Adj s w := by
  simp [hw]

/-- Adjacency in `G.replaceVertex s t` which does not involve `s` or `t` is the same as that of
`G`. -/
lemma adj_replaceVertex_iff_of_ne {v w : V} (hv : v ≠ t) (hw : w ≠ t) (hv' : v ≠ s) (hw' : w ≠ s) :
    (G.replaceVertex s t).Adj v w ↔ G.Adj v w := by
  simp [hv, hw, hv', hw']

```

**Review by ocfnash — APPROVED** [review]
Thanks!

Please apply the suggestions and then feel free to merge.

bors d+

**Comment by Parcly-Taxel** [comment]
bors merge

**Comment by Parcly-Taxel** [comment]
bors merge

**Pull request closed without merging**